From dc34af3e7673b9389927285c6ebc42a12c063ccb Mon Sep 17 00:00:00 2001
From: Ping Gao <ping.a.gao@intel.com>
Date: Fri, 2 Sep 2016 11:08:58 +0800
Subject: [PATCH 5/5] add a new parameter to define cap

add a new parameter vgt_cap to define the MAX GPU resource portion
that guest could take, the valid set is [0~100], default value '0'
means no upper cap.

Signed-off-by: Ping Gao <ping.a.gao@intel.com>
Reviewed-by: Kevin Tian <kevin.tian@intel.com>
Signed-off-by: Bing Niu <bing.niu@intel.com>
diff --git a/hw/display/vgt_vga.c b/hw/display/vgt_vga.c
index bff5a98..e66380c 100644
--- a/hw/display/vgt_vga.c
+++ b/hw/display/vgt_vga.c
@@ -74,6 +74,7 @@ int vgt_high_gm_sz = 448; /* in MB */
 int vgt_fence_sz = 4;
 int vgt_primary = 1; /* -1 means "not specified */
 int guest_domid = 0;
+int vgt_cap = 0;
 
 static int vgt_host_pci_cfg_get(VGTHostDevice *host_dev,
                                 void *data, int len, uint32_t addr);
@@ -99,9 +100,12 @@ static void create_vgt_instance(VGTVGAState *vdev)
     int domid = vdev->domid;
   
     qemu_log("vGT: %s: domid=%d, low_gm_sz=%dMB, high_gm_sz=%dMB, "
-        "fence_sz=%d, vgt_primary=%d\n", __func__, domid,
-        vgt_low_gm_sz, vgt_high_gm_sz, vgt_fence_sz, vgt_primary);
-    if (vgt_low_gm_sz <= 0 || vgt_high_gm_sz <=0 ||
+        "fence_sz=%d, vgt_primary=%d, vgt_cap=%d\n", __func__, domid,
+       	vgt_low_gm_sz, vgt_high_gm_sz, vgt_fence_sz, vgt_primary,
+      	vgt_cap);
+  
+ if (vgt_low_gm_sz <= 0 || vgt_high_gm_sz <=0 ||
+                vgt_cap < 0 || vgt_cap > 100 ||
 		vgt_primary < -1 || vgt_primary > 1 ||
         vgt_fence_sz <=0) {
         qemu_log("vGT: %s failed: invalid parameters!\n", __func__);
@@ -117,8 +121,9 @@ static void create_vgt_instance(VGTVGAState *vdev)
      * driver to create a vgt instanc for Domain domid with the required
      * parameters. NOTE: aperture_size and gm_size are in MB.
      */
-    if (!err && fprintf(vgt_file, "%d,%u,%u,%u,%d\n", domid,
-        vgt_low_gm_sz, vgt_high_gm_sz, vgt_fence_sz, vgt_primary) < 0) {
+    if (!err && fprintf(vgt_file, "%d,%u,%u,%u,%d,%u\n", domid,
+        vgt_low_gm_sz, vgt_high_gm_sz, vgt_fence_sz, vgt_primary,
+		vgt_cap) < 0) {
         err = errno;
     }
 
diff --git a/include/hw/i386/pc.h b/include/hw/i386/pc.h
index 8cde9c0..1ef55d2 100644
--- a/include/hw/i386/pc.h
+++ b/include/hw/i386/pc.h
@@ -332,6 +332,7 @@ int isa_vga_mm_init(hwaddr vram_base,
 extern int vgt_low_gm_sz;
 extern int vgt_high_gm_sz;
 extern int vgt_fence_sz;
+extern int vgt_cap;
 
 DeviceState *vgt_vga_init(PCIBus *pci_bus);
 void vgt_bridge_pci_conf_init(PCIDevice *dev);
diff --git a/qemu-options.hx b/qemu-options.hx
index 74199c9..19457d2 100644
--- a/qemu-options.hx
+++ b/qemu-options.hx
@@ -1265,6 +1265,15 @@ STEXI
 Start in full screen.
 ETEXI
 
+DEF("vgt_cap", 1, QEMU_OPTION_vgt_cap,
+    "-vgt_cap [num] allocated percentage for GPU resource portion\n",
+    QEMU_ARCH_ALL)
+STEXI
+@item -vgt_cap @var{size}
+@findex -vgt_cap
+Set the allocated GPU resource portion.
+ETEXI
+
 DEF("g", 1, QEMU_OPTION_g ,
     "-g WxH[xDEPTH]  Set the initial graphical resolution and depth\n",
     QEMU_ARCH_PPC | QEMU_ARCH_SPARC)
diff --git a/vl.c b/vl.c
index a104f9a..5ae1ca5 100644
--- a/vl.c
+++ b/vl.c
@@ -4073,6 +4073,12 @@ int main(int argc, char **argv, char **envp)
                     vgt_fence_sz = strtol(optarg, &ptr, 10);
                 }
                 break;
+            case QEMU_OPTION_vgt_cap:
+                {
+                   char *ptr;
+                   vgt_cap = strtol(optarg, &ptr, 10);
+                }
+		break;
 #endif
             default:
                 os_parse_cmd_args(popt->index, optarg);
